### DISCUSSION TITLE
브라우저가 동시성 문제를 어떻게 해결했는지 설명해 주세요

### DISCUSSION THREAD

**Discussion by hwangstar156** [woowacourse-study/2022-woowahan-bansanghwe#64] [JavaScript]
자바스크립트는 싱글 스레드 언어입니다. 하지만 브라우저는 여러 작업을 동시에 하는것 처럼 느껴집니다.

이는 브라우저의 동시성 때문입니다.

참고로 여기서 동시성은 실제 물리적으로 동시에 실행되는 병렬적인 작업이 아니라 하나의 흐름으로 실행하지만 그 흐름을 작은 단위로 나누어 실행시켜 동시에 일어나는것 처럼 보이게 하는것이다.

 그런데 자바스크립트는 싱글스레드 언어인데 어떻게 브라우저의 동시성을 지원하는걸까??

## 이벤트 루프

답은 바로 '이벤트 루프'이다. 이벤트 루프는 태스크가 들어오길 기다렸다가 태스크가 들어오면 이를 처리하고, 처리할 태스크가 없는 경우엔 잠드는, 끊임없이 돌아가는 자바스크립트 내 루프다.

이벤트 루프는 어떻게 이루어져있을까요?

크게 call Stack, web api, task queue 이렇게 나눌수 있고 task queue는 마이크로태스크큐, 매크로 태스크큐로 분리할수 있습니다.

그러면 각각의 요소가 하는 기능을 알아보겠습니다.

먼저 call stack입니다. 기본적으로 자바스크립트 코드가 실행이 되면 callStack에 쌓이게 됩니다. 

다음 web api 입니다. javascript가 아닌 브라우저에서 지원하는 api이다. ajax, setTimeout, Dom등의 api를 지원해줍니다.

다음은 태스크큐입니다. 태스크큐(메크로, 마이크로)는 현재 실행중인 태스크가 없을때(call stack)이 비었을때 태스크큐의 첫번째 태스크를 꺼내와서 실행합니다.

태스크 큐중 매크로 태스크큐는 setTimeout, setInterval, event등이 실행하면 매크로 태스크큐로 들어가게 됩니다.

매크로태스크큐는 특정태스크를 처리할때까지 절대 렌더링을 진행하지 않습니다. 그래서 태스크가 매우 길어진다면 태스크가 진행중일때 발생하는 사용자 이벤트를 처리할수가 없습니다.

다음은 마이크로 태스크큐입니다. 마이크로 태스크큐는 우선순위가 더 높은 태스크큐입니다. 자바스크립트 엔진은 매크로태스크 하나를 처리할 때마다 또 다른 매크로태스크나 렌더링 작업을 하기 전에 마이크로태스크 큐에 쌓인 마이크로태스크 전부를 처리합니다.

마이크로 태스크큐에 담기는 태스크는 Promise를 통해 만들고 then, catch, finally 핸들러, await 등에 해당합니다.

이벤트 루프 알고리즘은 

- 가장 오래된 매크로 태스크큐 실행
- 모든 마이크로 태스크큐 실행
- 렌더링 진행
- 다음 매크로 태스크큐 실행

이런식으로 진행됩니다.

async await vs promise 둘다 마이크로 태스크큐에 들어가게 되지만 약간의 동작방식이 다릅니다. promise의 경우에는 마이크로 태스크큐에 넣은후 그다음 코드를 바로 콜스택에 넣게 됩니다. 하지만 async await 같은 경우에는 await를 만나게 되면 

브라우저는 왜 동시성을 지원할까??

**Comment by jswith** [comment]
동시성이 무엇인가요?

**Comment by jswith** [comment]
브라우저에서는 왜 동시성이라는 개념을 중시할까요?

**Comment by jswith** [comment]
왜 브라우저는 두 가지일을 병렬적으로 할 수 없는건가요?

**Comment by jswith** [comment]
브라우저도 싱글 스레드인가요?

**Comment by jswith** [comment]
이벤트 루프에 대해서 간략하게 설명해주세요

**Comment by jswith** [comment]
이벤트 루프는 브라우저의 어느 영역에서 동작하나요?

**Comment by jswith** [comment]
비동기가 이루어지는 일련의 과정을 설명해주세요

**Comment by jswith** [comment]
스레드의 개념에 대해서 말씀해주세요

**Comment by jswith** [comment]
이벤트 루프에서 비동기 외 함수는 바로 처리하나요?

**Comment by jswith** [comment]
콜 스택이 계속 차있을 경우, 태스크 큐에 들어간 작업들이 무한정 대기할 수 있을 것 같은데 어떻게 처리하나요?

**Comment by jswith** [comment]
콜 스택이 터지는 경우 어떤 에러가 발생하나요?

**Comment by jswith** [comment]
콜 스택이 터지는 에러를 경험해보신 적이 있나요?

**Comment by jswith** [comment]
재귀가 무엇인가요?

**Comment by jswith** [comment]
맥시멈 콜 스택 에러가 발생했던 이유?

**Comment by jswith** [comment]
재귀를 끝내는 방법?

**Comment by jswith** [comment]
setTimeout을 사용한 후에 clearTimeout을 사용해야 하는 이유?

**Comment by jswith** [comment]
시간이 다 끝났다면 clearTimeout을 사용할 필요가 없나요?

**Comment by jswith** [comment]
Promise에 대해서 설명해주세요

**Comment by jswith** [comment]
Promise의 내부 동작 방식에 대해서 설명해주세요

**Comment by jswith** [comment]
async/await에 대해서 설명해주세요

**Comment by jswith** [comment]
마이크로 태스크큐와 매크로 태스크큐의 차이

**Comment by jswith** [comment]
마이크로 태스크큐와 매크로 태스크큐를 왜 나누어 둔걸까요?

**Comment by jswith** [comment]
Promise vs async/await 코드 내부의 차이

**Comment by jswith** [comment]
Generator가 어떤 개념을 사용해서 코드를 기다리게 할 수 있는지?

**Comment by jswith** [comment]
Generator의 동작 방식

**Comment by jswith** [comment]
Generator의 마지막 실행에서도 next가 있는지?

**Comment by jswith** [comment]
Generator에서 done 말고 다른 상태도 있는지

**Comment by jswith** [comment]
자바스크립트에서 fetch를 실행하고 난 이후의 동작?

**Comment by jswith** [comment]
비동기 통신에서 라이브러리를 사용한 경험?

**Comment by jswith** [comment]
리액트 쿼리를 사용한 이유?

**Comment by jswith** [comment]
리액트 쿼리에서 캐싱을 어떻게 사용했나요?

**Comment by jswith** [comment]
자바스크립트는 왜 싱글 스레드를 선택했을까?

**Comment by jswith** [comment]
이벤트 루프는 어디에 존재하는가?
